lib/TPP/Dialect/VNNI/TransformOps/VNNITransformOps.cpp

```diff
@@ -0,0 +1,102 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Dialect/VNNI/TransformOps/VNNITransformOps.h"
+#include "TPP/Dialect/VNNI/VNNIOps.h"
+#include "TPP/Transforms.h"
+#include "mlir/AsmParser/AsmParser.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Transform/IR/TransformDialect.h"
+#include "mlir/Dialect/Transform/IR/TransformInterfaces.h"
+#include "mlir/Interfaces/ViewLikeInterface.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Casting.h"
+#include "TPP/Dialect/Tpp/TppOps.h"
+
+using namespace mlir;
+using namespace mlir::transform;
+using namespace mlir::tpp;
+//===----------------------------------------------------------------------===//
+// Transform op registration
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+class VNNITransformDialectExtension
+    : public transform::TransformDialectExtension<
+          VNNITransformDialectExtension> {
+public:
+  using Base::Base;
+
+  void init() {
+    registerTransformOps<
+#define GET_OP_LIST
+#include "TPP/Dialect/VNNI/TransformOps/VNNITransformOps.cpp.inc"
+        >();
+  }
+};
+} // namespace
+
+namespace {
+// A simple pattern rewriter that implements no special logic.
+class SimpleRewriter : public PatternRewriter {
+public:
+  SimpleRewriter(MLIRContext *context) : PatternRewriter(context) {}
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// MapAndConvertLinalgToTpp
+//===----------------------------------------------------------------------===//
+
+// Convert a vnni.matmul to a tpp.vnni_matmul.
+struct MapVNNIMatmulToTpp : public OpRewritePattern<vnni::MatmulOp> {
+  using OpRewritePattern<vnni::MatmulOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vnni::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    if (!matmulOp.hasBufferSemantics())
+      return rewriter.notifyMatchFailure(
+          matmulOp, "Expect buffer semantics when mapping to tpp");
+    if (matmulOp.hasDynamicShape())
+      return rewriter.notifyMatchFailure(
+          matmulOp, "Expect static shape when mapping to tpp");
+    rewriter.replaceOpWithNewOp<tpp::VNNIMatmulOp>(
+        matmulOp, matmulOp.getMatrixA(), matmulOp.getMatrixB(),
+        matmulOp.getMatrixC());
+    return success();
+  }
+};
+
+  
+DiagnosedSilenceableFailure transform::MapVNNIToTppOp::applyToOne(
+    Operation *target, SmallVector<Operation *> &results,
+    TransformState &state) {
+  if (!target->hasTrait<OpTrait::IsIsolatedFromAbove>()) {
```

Comment:
can you please document this in the op description? It also seems that the patter in this case is relatively easy so we can get away with a failureOr<tpp::VNNIMatmul> instead of running the pattern rewriter?

Reply:
Running the conversion on an entire function requires that I use pattern rewriter, because applyToOne does not get invoked with just one op (vnni.matmul) but with the whole function, right? Also, I am invoking conversion on the whole function because we need to convert brgemm calls as well. Does this seem right to you?

Reply:
I have updated the patch so that "map_vnni_to_tpp" transform applies to vnni.matmul ops directly instead of func. With this change, the op needn't be isolated from above and hence, I have removed it.

Reply:
I think for the VNNI to TPP pass it should be fine to be a function pass because it will always be 1:1 and the shapes should have been checked (and matched) by the Linalg to VNNI pass.